File: r/triangle/triangle.R
triangle <- function(x, y, z) {
  t <- sort(c(x, y, z))
  if (t[1] <= 0 || t[1] + t[2] <= t[3]) {
    stop("not a triangle")
  }
  class(t) <- append(class(t),
    if (t[1] == t[2] && t[1] == t[3]) {
      c("equilateral", "isosceles")
    } else if (t[1] == t[2] || t[2] == t[3]) {
      "isosceles"
    } else {
      "scalene"
    }
  )
  t
}

File: julia/triangle/triangle.jl
module Triangle
    export is_equilateral, is_isosceles, is_scalene

    @enum Classification begin
        invalid
        equilateral
        isosceles
        scalene
    end

    function is_valid(sides)
        valid = false
        if length(sides) == 3
            a, b, c = sort(sides)
            valid = a > 0 && a + b > c
        end
        valid
    end

    function classification(sides)
        cls = invalid
        if is_valid(sides)
            set = Set(sides)
            len = length(set)

            if len == 1
                cls = equilateral
            elseif len == 2
                cls = isosceles
            else
                cls = scalene
            end
        end
        cls
    end

    function is_equilateral(sides)
        classification(sides) == equilateral
    end

    function is_isosceles(sides)
        in(classification(sides), [equilateral, isosceles])
    end

    function is_scalene(sides)
        classification(sides) == scalene
    end
end

using .Triangle
